### ISSUE TITLE
Issues with scratch images after 3.0.4 (3.1.0)

### ISSUE THREAD

**Issue by schwichtgit** [docker/for-mac#5405]
For some time I have been creating base centos7 images with this Dockerfile:

```
FROM scratch
ARG pg_version
ADD dl/${pg_version}/centos-7-docker.tar.xz /
COPY files/variants/${pg_version}/etc/yum.repos.d/*repo /etc/yum.repos.d/

RUN rm -f /etc/yum.repos.d/CentOS* && \
    rpm --import  /etc/pki/rpm-gpg/RPM-GPG-KEY-CentOS-7
LABEL   org.label-schema.schema-version="1.0" \
        org.label-schema.vendor="CentOS" \
        org.label-schema.license="GPLv2"
CMD ["/usr/bin/bash"]
```

This uses the ADD command to extract the centos-7-docker.tar.xz tar ball.

Docker Desktop Community for Mac 3.1.0 fails with an error that /bin/sh
can not be found. 
Commenting
```
RUN rm -f /etc/yum.repos.d/CentOS* && \
    rpm --import  /etc/pki/rpm-gpg/RPM-GPG-KEY-CentOS-7
```
allows to create an image

if I install version 3.0.4 the identical scenario works fine.
```
Step 1/10 : FROM scratch as intermediate
 --->
Step 2/10 : ARG pg_version
 ---> Running in e010df2c4bdc
Removing intermediate container e010df2c4bdc
 ---> 683672f2d60f
Step 3/10 : ADD dl/${pg_version}/centos-7-docker.tar.xz /
 ---> 6a66f17355b2
Step 4/10 : COPY files/variants/${pg_version}/etc/yum.repos.d/*repo /etc/yum.repos.d/
 ---> c0744d93f241
Step 5/10 : RUN rm -f /etc/yum.repos.d/CentOS* &&     rpm --import  /etc/pki/rpm-gpg/RPM-GPG-KEY-CentOS-7
 ---> Running in 3b9da6c1145f
Removing intermediate container 3b9da6c1145f
 ---> 7537b54aa4cb
Step 6/10 : LABEL   org.label-schema.schema-version="1.0"         org.label-schema.vendor="CentOS"         org.label-schema.license="GPLv2"
 ---> Running in a7f500c11666
Removing intermediate container a7f500c11666
 ---> 6ddb7f3b1501
Step 7/10 : CMD ["/usr/bin/bash"]
 ---> Running in 97d5768fad9b
Removing intermediate container 97d5768fad9b
 ---> ffee6463e8ae
Step 8/10 : LABEL build_date=20210301173219
 ---> Running in 469d57cf76e6
Removing intermediate container 469d57cf76e6
 ---> d3c5bd5d17c9
Step 9/10 : LABEL org.centos=centos7.9.2009
 ---> Running in fd639cbb87b5
Removing intermediate container fd639cbb87b5
 ---> daca5a3009c9
Step 10/10 : LABEL org.label-schema.name=centos7.9.2009 Base Image
 ---> Running in 949b11114271
Removing intermediate container 949b11114271
 ---> 8fc2de19520e
Successfully built 8fc2de19520e
```

however on 3.1.0:
```
Sending build context to Docker daemon  230.1MB
Step 1/10 : FROM scratch as intermediate
 --->
Step 2/10 : ARG pg_version
 ---> Running in 2b0ef01b487e
Removing intermediate container 2b0ef01b487e
 ---> 603281d7d84f
Step 3/10 : ADD dl/${pg_version}/centos-7-docker.tar.xz /
 ---> d39b3585f9f2
Step 4/10 : COPY files/variants/${pg_version}/etc/yum.repos.d/*repo /etc/yum.repos.d/
 ---> 5d01451d9825
Step 5/10 : RUN rm -f /etc/yum.repos.d/CentOS* &&     rpm --import  /etc/pki/rpm-gpg/RPM-GPG-KEY-CentOS-7
 ---> Running in a77e91de4960
OCI runtime create failed: container_linux.go:370: starting container process caused: exec: "/bin/sh": stat /bin/sh: no such file or directory: unknown
make[1]: *** [build] Error 1
make: *** [os] Error 2
```

I create the tar ball with this Makefile (not important, so just FYI):
```
iso7:=/install/mirror/centos7/7.6.1810/os/x86_64/images/boot.iso
iso7:=/install/mirror/centos7/7.9.2009/os/x86_64/images/boot.iso
iso8:=/install/mirror/centos8/8.3.2011/BaseOS/x86_64/kickstart/images/boot.iso
ks7:=/home/work/projects/sig-cloud-instance-build/docker/centos-7-x86_64.ks
ks8:=/home/work/projects/sig-cloud-instance-build/docker/centos-8.ks
tmp:=./tmp/
kernel_args := ksdevice=eth0 net.ifnames=0 biosdevname=0

all:

centos7:
	sudo livemedia-creator --make-tar --iso $(iso7) --tmp $(tmp) --ks $(ks7) --kernel-args '$(kernel_args)' --image-name centos-7-docker.tar.xz
centos8:
	sudo livemedia-creator --make-tar --iso $(iso8) --tmp $(tmp) --ks $(ks8) --kernel-args '$(kernel_args)' --image-name centos-8-docker.tar.xz


```

  - [x] I have tried with the latest version of Docker Desktop
  - [ ] I have tried disabling enabled experimental features
  - [x] I have uploaded Diagnostics
  - Diagnostics ID: 3EB5269B-F446-4400-94CC-917CAE8558ED/20210301173848

### Expected behavior
see the build using docker 3.0.4
### Actual behavior
see the build using docker 3.1.0
### Information

  - macOS Version: Big Sur 11.2.1
  - Docker Desktop Version: 3.1.0

### Steps to reproduce the behavior
see above

**Comment by thaJeztah** [comment]
That's really weird that it worked on 3.0.4; the `FROM scratch` image is literally an empty image (nothing in the image, so no shell or other binaries), so that step _failing_ is expected (the unexpected part is that for some reason it "worked" in your old install, because it shouldn't)

Testing a new build (both tried on current docker 20.10 engine, and on an older 19.03 engine);

```bash
mkdir /test && cd test
touch some-file.txt

cat > Dockerfile <<EOF
FROM scratch
COPY some-file.txt .
RUN rm -f some-file.txt
EOF

docker build .
```

Running the above;

```bash
docker build .
Sending build context to Docker daemon   2.56kB
Step 1/3 : FROM scratch
 --->
Step 2/3 : COPY some-file.txt .
 ---> 94fd5ef8a6e5
Step 3/3 : RUN rm -f some-file.txt
 ---> Running in c934eb98bf6b
OCI runtime create failed: container_linux.go:349: starting container process caused "exec: \"/bin/sh\": stat /bin/sh: no such file or directory": unknown
```

The only situation I can think of is, if for some reason the failed build was committed to an image, then the "classic" (non-buildkit) builder could use that image for caching (the classic builder is somewhat naive in it's caching mechanism, and relies on image metadata in the local image cache as "source of truth")

For example, to illustrate that situation, build the image with `--rm=false` option, then manually commit the failed container to an image;

```bash
$ docker build --rm=false .
Sending build context to Docker daemon   2.56kB
Step 1/3 : FROM scratch
--->
Step 2/3 : COPY some-file.txt .
---> Using cache
---> 94fd5ef8a6e5
Step 3/3 : RUN rm -f some-file.txt
---> Running in e49b0cd0db85
OCI runtime create failed: container_linux.go:349: starting container process caused "exec: \"/bin/sh\": stat /bin/sh: no such file or directory": unknown
````

The build failed, but (because of the `--rm=false` option), the failed container is still present. Now commit the failed container (with ID `e49b0cd0db85` from the output above) to an image;

```bash
$ docker commit e49b0cd0db85 temp-image
sha256:d4e52a1a4ee97689254611b9704e44cc8d75b6c71c0f267ec79753a6ec59b4f4
```

After this, `docker build` (with the classic builder) will match that image, and _assume_ it was a succesful build, therefore use it as cache for the `RUN` step;

```bash
$ docker build --rm=false .
Sending build context to Docker daemon   2.56kB
Step 1/3 : FROM scratch
--->
Step 2/3 : COPY some-file.txt .
---> Using cache
---> 94fd5ef8a6e5
Step 3/3 : RUN rm -f some-file.txt
---> Using cache
---> d4e52a1a4ee9
Successfully built d4e52a1a4ee9
```

The only other situation I could think of is if a stage named `scratch` was created in a previous stage in a multi-stage Dockerfile.

For example, creating a "scratch" stage;

```bash
cat > Dockerfile <<EOF
FROM busybox AS scratch

FROM scratch
COPY some-file.txt .
RUN rm -f some-file.txt
EOF
```

```bash
$ docker build .
Sending build context to Docker daemon   2.56kB
Step 1/4 : FROM busybox AS scratch
---> 491198851f0c
Step 2/4 : FROM scratch
---> 491198851f0c
Step 3/4 : COPY some-file.txt .
---> cd918dfb82bd
Step 4/4 : RUN rm -f some-file.txt
---> Running in 6b13c3d8e1b7
Removing intermediate container 6b13c3d8e1b7
---> 2529583be066
Successfully built 2529583be066
```

I also tried committing an _image_ as scratch, but I think there's special handling for the `FROM scratch` case, which won't use an image for `scratch`;

```bash
$ docker pull busybox
$ docker tag busybox scratch
```

```bash
$ docker build --no-cache .
Sending build context to Docker daemon   2.56kB
Step 1/3 : FROM scratch
 --->
Step 2/3 : COPY some-file.txt .
 ---> 5af65f8412d0
Step 3/3 : RUN rm -f some-file.txt
 ---> Running in ac9ad668e033
OCI runtime create failed: container_linux.go:349: starting container process caused "exec: \"/bin/sh\": stat /bin/sh: no such file or directory": unknown
```


The only other situation I could think of is if it would be a multi-stage Dockerfile, and a stage named `scratch` was created in a previous stage;

For example, creating a "scratch" stage;

```bash
cat > Dockerfile <<EOF
FROM busybox AS scratch

FROM scratch
COPY some-file.txt .
RUN rm -f some-file.txt
EOF
```

```bash
$ docker build .
Sending build context to Docker daemon   2.56kB
Step 1/4 : FROM busybox AS scratch
---> 491198851f0c
Step 2/4 : FROM scratch
---> 491198851f0c
Step 3/4 : COPY some-file.txt .
---> cd918dfb82bd
Step 4/4 : RUN rm -f some-file.txt
---> Running in 6b13c3d8e1b7
Removing intermediate container 6b13c3d8e1b7
---> 2529583be066
Successfully built 2529583be066
```

**Comment by thaJeztah** [comment]
Hm.. never mind my previous comment, so I "understand" why it worked before, and may have a clue why it may not work in 3.1.0; the `.xz` file you're adding contains a full rootfs for the image, so after `ADD`-ing the file, it would be decompressed (and now the image would have `/bin/sh`, so commands are able to run).

I recall that some versions of Docker Desktop did not have support for `xz` archives, in which case the rootfs would not be extracted (possibly "silently", I'd have to check), so the image would only contain the compressed archive.

I recall this was fixed recently, but I don't know if this already shipped; let me check that with the Docker Desktop team

**Comment by thaJeztah** [comment]
Confirmed; I think that's the problem.


Here's in a `docker:20.10-dind` container;

Create a `.xz` archive to test with;

```bash
mkdir test && cd test
touch one two three
tar -cJ -f archive.xz one two three
```

Create a Dockerfile that `ADD`'s the archive in a `/foo/` directory, and after that lists the content of the directory:

```bash
cat > Dockerfile <<EOF
FROM busybox
WORKDIR /foo
ADD archive.xz /foo/
RUN ls -la /foo/
EOF
```

Building (with `xz` present);

```bash
$ docker build --no-cache .
Sending build context to Docker daemon  4.608kB
Step 1/4 : FROM busybox
latest: Pulling from library/busybox
5c4213be9af9: Pull complete
Digest: sha256:c6b45a95f932202dbb27c31333c4789f45184a744060f6e569cc9d2bf1b9ad6f
Status: Downloaded newer image for busybox:latest
---> 491198851f0c
Step 2/4 : WORKDIR /foo
---> Running in 81e9a22f4b05
Removing intermediate container 81e9a22f4b05
---> 3fc7c6c027cf
Step 3/4 : ADD archive.xz /foo/
---> 37377e1d8373
Step 4/4 : RUN ls -la /foo/
---> Running in 55374719d30d
total 8
drwxr-xr-x    1 root     root          4096 Mar  2 12:11 .
drwxr-xr-x    1 root     root          4096 Mar  2 12:11 ..
-rw-r--r--    1 root     root             0 Mar  2 12:09 one
-rw-r--r--    1 root     root             0 Mar  2 12:09 three
-rw-r--r--    1 root     root             0 Mar  2 12:09 two
Removing intermediate container 55374719d30d
---> 41fb8623e447
Successfully built 41fb8623e447
```

Now "uninstall" `xz`:

```bash
rm /usr/bin/xz
```

And repeat the build:

```bash
$ docker build --no-cache .
Sending build context to Docker daemon  4.608kB
Step 1/4 : FROM busybox
---> 491198851f0c
Step 2/4 : WORKDIR /foo
---> Running in 9e111979f492
Removing intermediate container 9e111979f492
---> 71a7db2edde0
Step 3/4 : ADD archive.xz /foo/
---> 80842ecbaf7a
Step 4/4 : RUN ls -la /foo/
---> Running in 0e7bfb3f04aa
total 12
drwxr-xr-x    1 root     root          4096 Mar  2 12:12 .
drwxr-xr-x    1 root     root          4096 Mar  2 12:12 ..
-rw-r--r--    1 root     root           160 Mar  2 12:10 archive.xz
Removing intermediate container 0e7bfb3f04aa
---> adf907e4010f
Successfully built adf907e4010f
```

This time, the archive is not extracted, so it's added as-is (no warning is printed)

**Comment by thaJeztah** [comment]
I see the fix for this was merged 19 days ago in Docker Desktop (internal issue [DESKTOP-3601]). Asking if it has shipped already

**Comment by stephen-turner** [comment]
Ah yes, thank you @thaJeztah. This was released in 3.2.0, but we pulled that release because of other problems, so it will be in 3.2.1 later this week.

**Comment by stephen-turner** [comment]
Public bug reference is #5271.

**Comment by schwichtgit** [comment]
I just installed Docker Desktop/mac 3.2.2(61853)

and confirm the add with XZ compressed archives works again:
```
docker build --no-cache --rm=true --file=variants/centos7/Dockerfile.os  \
		  --build-arg pg_version="centos7.9.2009" \
		  --label "build_date=20210317130508" \
		  --label "org.centos=centos7.9.2009" \
		  --label "org.label-schema.name=centos7.9.2009 Base Image" \
		  --tag xxxxxx/centos:centos7.9.2009 .
#1 [internal] load build definition from Dockerfile.os
#1 sha256:1ff863dfe386db0be40c930300cb5706d84a10c567c8d0470aa3015029869dd6
#1 transferring dockerfile: 463B 0.0s done
#1 DONE 0.0s

#2 [internal] load .dockerignore
#2 sha256:bc556302b4a58b4f83e3c78b5b59fae2f1bce4ce4c135879e43f40a6a6901e3b
#2 transferring context: 2B done
#2 DONE 0.0s

#3 [internal] load build context
#3 sha256:b27a5a04329375c67face57b7087720b8d6dce6ee8ab97b5d77bb47c4f8638dc
#3 transferring context: 45.83MB 1.0s done
#3 DONE 1.1s

#4 [1/3] ADD dl/centos7.9.2009/centos-7-docker.tar.xz /
#4 sha256:75311090107f1fa7e68c0cc17d706fbb8e760b0aeba41f5ff2ba0f747bd053a7
#4 DONE 5.1s

#5 [2/3] COPY files/variants/centos7.9.2009/etc/yum.repos.d/*repo /etc/yum.repos.d/
#5 sha256:61be4f34dd1e8b0f6511f48d06875a295dbafba8c3086bab8e9c2c3bb5b2b263
#5 DONE 0.7s

#6 [3/3] RUN rm -f /etc/yum.repos.d/CentOS* &&     rpm --import  /etc/pki/rpm-gpg/RPM-GPG-KEY-CentOS-7
#6 sha256:10f6e8d93da921dab529a466e6a2e661de7e94acf0d0750ddaefbc1bb8f870f6
#6 DONE 0.6s

#7 exporting to image
#7 sha256:e8c613e07b0b7ff33893b694f7759a10d42e180f2b4dc349fb57dc6b71dcab00
#7 exporting layers
#7 exporting layers 1.8s done
#7 writing image sha256:e0ae008730d086f6ba0bff02752e7aa1e1f078d6c2a80c8179eea8c1984ce3af done
#7 naming to docker.io/xxxxxx/centos:centos7.9.2009 done
#7 DONE 1.8s
```

**Comment by schwichtgit** [comment]
closing this issue as resolved in  3.2.2(61853)
